src/ast/values.ts

```diff
@@ -12,3 +32,229 @@ export const UNKNOWN_EXPRESSION: ExpressionEntity = {
+		hasEffectsWhenCalledAtPath: path => {
+			if (path.length === 1) {
+				const subPath = path[0];
+				return isUnknownKey(subPath) || !literalMembers[type][subPath];
+			}
+			return true;
+		},
+		someReturnExpressionWhenCalledAtPath: (
+			path: ObjectPath,
+			_callOptions: CallOptions,
+			predicateFunction: SomeReturnExpressionCallback,
+			options: ExecutionPathOptions
+		) => {
+			if (path.length === 1) {
+				const subPath = path[0];
+				return isUnknownKey(subPath)
+					|| !literalMembers[type][subPath]
+					|| predicateFunction(options)(literalMembers[type][subPath].returns);
+			}
+			return true;
+		},
+		toString: () => '[[UNKNOWN STRING]]'
+	};
+}
+
+export const UNKNOWN_LITERALS = {
+	boolean: createUnknownLiteral('boolean'),
+	number: createUnknownLiteral('number'),
+	string: createUnknownLiteral('string')
+};
+const returnsBoolean: RawMemberDescription = { value: { returns: UNKNOWN_LITERALS.boolean, callsArgs: null } };
+const callsArgReturnsBoolean: RawMemberDescription = { value: { returns: UNKNOWN_LITERALS.boolean, callsArgs: [0] } };
+const returnsNumber: RawMemberDescription = { value: { returns: UNKNOWN_LITERALS.number, callsArgs: null } };
+const callsArgReturnsNumber: RawMemberDescription = { value: { returns: UNKNOWN_LITERALS.number, callsArgs: [0] } };
+const returnsString: RawMemberDescription = { value: { returns: UNKNOWN_LITERALS.string, callsArgs: null } };
+
+export const UNKNOWN_OBJECT_EXPRESSION: ExpressionEntity = {
+	reassignPath: () => {},
+	forEachReturnExpressionWhenCalledAtPath: () => {},
+	getValue: () => UNKNOWN_VALUE,
+	hasEffectsWhenAccessedAtPath: path => path.length > 1,
+	hasEffectsWhenAssignedAtPath: path => path.length > 1,
+	hasEffectsWhenCalledAtPath: path => {
+		if (path.length === 1) {
+			const subPath = path[0];
+			return isUnknownKey(subPath) || !objectMembers[subPath];
+		}
+		return true;
+	},
+	someReturnExpressionWhenCalledAtPath: (
+		path: ObjectPath,
+		_callOptions: CallOptions,
+		predicateFunction: SomeReturnExpressionCallback,
+		options: ExecutionPathOptions
+	) => {
+		if (path.length === 1) {
+			const subPath = path[0];
+			return isUnknownKey(subPath)
+				|| !objectMembers[subPath]
+				|| predicateFunction(options)(objectMembers[subPath].returns);
+		}
+		return true;
+	},
+	toString: () => '[[UNKNOWN OBJECT]]'
+};
+
+export const arrayMembers: MemberDescriptions =
+	assembleMemberDescriptions({
+		concat: returnsArray,
+		every: callsArgReturnsBoolean,
+		filter: callsArgReturnsArray,
+		find: callsArgReturnsUnknown,
+		findIndex: callsArgReturnsNumber,
+		forEach: callsArgReturnsUnknown,
+		includes: returnsBoolean,
+		indexOf: returnsNumber,
+		join: returnsString,
+		lastIndexOf: returnsNumber,
+		map: callsArgReturnsArray,
+		reduce: callsArgReturnsUnknown,
+		reduceRight: callsArgReturnsUnknown,
+		slice: returnsArray,
+		some: callsArgReturnsBoolean,
+		toLocaleString: returnsString,
+		toString: returnsString
```

Comment:
Is the assumption that these can work because we know we specifically have an array of literal values so there are no `toString` side effects?

If these cases work, surely `join` would work too then?

But if we don't actually know what we are calling `toString` on, I would argue we should remove these.

Reply:
Ah, I see, did not consider this meta-effect. So I guess yes, we could consider removing it. Which in the case of `join` would really be a pity. Or we act under the assumption that `toString` never has a side-effect. Might be reasonable to do as we also assume that builtins have not been modified. Even if a `toString` method is provided, I would assume that people would not rely on it having side-effects. I may be proven wrong but unless you know a specific package that relies on this, I would argue to maintain this assumption until someone encounters an issue here.
What do you think?

Reply:
`toString` should always be pure definitely, even if it is overridden. So it's the edge case of the edge case that catches yes... and we have assumptions like this all over for builtins so I suppose it is consistent with the assumptions of the project. It may be worth documenting somewhere these kinds of assumptions so that users can easily determine if Rollup is right for their codebase before hitting odd bugs like these though. Personally I find projects that clearly state their assumptions incredibly reassuring as you can fully appreciate the decisions you are making with it.